_posts/2021-04-20-custom-errors.md

```diff
@@ -0,0 +1,82 @@
+---
+layout: post
+published: true
+title:  Custom Errors in Solidity
+date:   2021-04-20
+author: Solidity Team
+category: Explainers
+---
+
+[Solidity v0.8.4](https://github.com/ethereum/solidity/releases/tag/v0.8.4) provides a convenient and gas-efficient way to explain to user why an operation failed. They can be defined inside and outside of contracts (including interfaces and libraries). 
+
+The syntax of Errors are similar to that of
+[Events](https://docs.soliditylang.org/en/latest/contracts.html#events). They have to be used
+together with the [revert
+statement](https://docs.soliditylang.org/en/latest/control-structures.html#revert-statement) which
+causes all changes in the current call to be reverted and passes the error data back to the caller.
+
+The following contract shows an example usage of an Error:
+
+```solidity
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity ^0.8.4;
+
+contract VendingMachine {
+    address payable owner = payable(msg.sender);
+    error Unauthorized();
+    function withdraw() public {
+        if (msg.sender != owner)
+            revert Unauthorized();
+
+        owner.transfer(address(this).balance);
+    }
+    // ...
+}
+```
+
+In the above example, `revert Unauthorized();` is equivalent to the following Yul code:
```

Comment:
Isn't this going a bit too deep?

Reply:
Maybe we should rather highlight the cost difference compared to `revert("Unauthorized")`?

Reply:
I wanted to include something that makes it clear that solidity isn't randomly assigning numbers to errors. Perhaps I can rephrase this slightly better, but still include the Yul code. 

I'll also highlight the cost difference.

Reply:
If it's not too complicated, maybe we should really just show how to decode it using ethers.js - in that way the encoding should be clear, shouldn't it?

Reply:
People will stop reading when they see "depth" and inline assembly. We can use this, but then it should be put at the very end of the blog post.